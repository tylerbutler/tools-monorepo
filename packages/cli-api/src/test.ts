// biome-ignore lint/performance/noBarrelFile: <explanation>
export { testWithTempDir } from "./tempDir.js";
