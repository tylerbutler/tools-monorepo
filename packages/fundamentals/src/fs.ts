import { access } from "node:fs/promises";

/**
 * Check if a file or directory exists.
 *
 * @remarks
 * This is an async replacement for the deprecated `fs.exists()` and synchronous `fs.existsSync()`.
 * It uses `fs.promises.access()` internally to check file accessibility.
 *
 * @param path - The path to check for existence
 * @returns A promise that resolves to `true` if the path exists, `false` otherwise
 *
 * @example
 * ```typescript
 * if (await exists('./myfile.txt')) {
 *   console.log('File exists');
 * }
 * ```
 *
 * @public
 */
export async function exists(path: string): Promise<boolean> {
	try {
		await access(path);
		return true;
	} catch {
		return false;
	}
}
