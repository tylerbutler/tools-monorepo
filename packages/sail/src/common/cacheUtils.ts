import type { Logger } from "@tylerbu/cli-api";
import type { SharedCacheManager } from "../core/sharedCache/index.js";

/**
 * Logger interface with error method for CLI commands.
 */
export interface LoggerWithError extends Logger {
	error(message: string, options: { exit: number }): never;
}

/**
 * Display cache statistics in a formatted way.
 */
export async function displayCacheStatistics(
	logger: Logger,
	sharedCache: SharedCacheManager | null,
): Promise<void> {
	if (!sharedCache) {
		return;
	}

	const stats = await sharedCache.getStatistics();

	logger.log("\n=== Cache Statistics ===");
	logger.log(`Total Entries: ${stats.totalEntries}`);
	logger.log(`Total Size: ${(stats.totalSize / 1024 / 1024).toFixed(2)} MB`);
	logger.log(`Cache Hits: ${stats.hitCount}`);
	logger.log(`Cache Misses: ${stats.missCount}`);

	if (stats.hitCount > 0 || stats.missCount > 0) {
		const total = stats.hitCount + stats.missCount;
		const hitRate = ((stats.hitCount / total) * 100).toFixed(1);
		logger.log(`Hit Rate: ${hitRate}%`);
	}

	if (stats.hitCount > 0) {
		logger.log(`Avg Restore Time: ${stats.avgRestoreTime.toFixed(1)}ms`);

		const timeSavedSec = stats.timeSavedMs / 1000;
		if (timeSavedSec > 60) {
			const minutes = Math.floor(timeSavedSec / 60);
			const seconds = (timeSavedSec % 60).toFixed(1);
			logger.log(`Time Saved: ${minutes}m ${seconds}s`);
		} else {
			logger.log(`Time Saved: ${timeSavedSec.toFixed(1)}s`);
		}
	}

	logger.log("========================\n");
}

/**
 * Display minimal cache statistics (just entries and size).
 */
export async function displayMinimalCacheStatistics(
	logger: Logger,
	sharedCache: SharedCacheManager | null,
): Promise<void> {
	if (!sharedCache) {
		return;
	}

	const stats = await sharedCache.getStatistics();

	logger.log("\n=== Cache Statistics ===");
	logger.log(`Total Entries: ${stats.totalEntries}`);
	logger.log(`Total Size: ${(stats.totalSize / 1024 / 1024).toFixed(2)} MB`);
	logger.log("========================\n");
}

/**
 * Validate that a cache directory is specified.
 */
export function validateCacheDir(
	logger: LoggerWithError,
	cacheDir: string | undefined,
): asserts cacheDir is string {
	if (!cacheDir) {
		logger.error(
			"No cache directory specified. Use --cache-dir or set SAIL_CACHE_DIR environment variable.",
			{ exit: 1 },
		);
	}
}

/**
 * Initialize the shared cache and handle errors.
 */
export async function initializeCacheOrFail(
	logger: LoggerWithError,
	cacheDir: string,
	skipCacheWrite = true,
	verifyCacheIntegrity = false,
): Promise<SharedCacheManager> {
	const { initializeSharedCache } = await import(
		"../core/sharedCache/index.js"
	);

	const sharedCache = await initializeSharedCache(
		cacheDir,
		process.cwd(),
		skipCacheWrite,
		verifyCacheIntegrity,
	);

	if (!sharedCache) {
		logger.error("Failed to initialize cache.", { exit: 1 });
	}

	return sharedCache;
}
