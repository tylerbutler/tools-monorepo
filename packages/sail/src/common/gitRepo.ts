import { exec } from "./utils.js";

const LINE_BREAK_REGEX = /\r?\n/;
const DELETED_FILE_REGEX = /^.?D /;

/**
 * @deprecated Should not be used outside the build-tools package.
 */
export class GitRepo {
	public constructor(public readonly resolvedRoot: string) {}

	public async getCurrentSha() {
		const result = await this.exec("rev-parse HEAD", "get current sha");
		return result.split(LINE_BREAK_REGEX)[0];
	}

	/**
	 * Returns a set containing repo root-relative paths to files that are deleted in the working tree.
	 */
	public async getDeletedFiles(): Promise<Set<string>> {
		const results = await this.exec("status --porcelain", "get deleted files");
		const allStatus = results.split("\n");
		// Deleted files are marked with D in the first (staged) or second (unstaged) column.
		// If a file is deleted in staged and then revived in unstaged, it will have two entries.
		// The first entry will be "D " and the second entry will be "??". Look for unstaged
		// files and remove them from deleted set.
		const deletedFiles = new Set(
			allStatus
				.filter((t) => t.match(DELETED_FILE_REGEX))
				.map((t) => t.substring(3)),
		);
		const untrackedFiles = allStatus
			.filter((t) => t.startsWith("??"))
			.map((t) => t.substring(3));
		for (const untrackedFile of untrackedFiles) {
			deletedFiles.delete(untrackedFile);
		}
		return deletedFiles;
	}

	/**
	 * Returns an array containing repo repo-relative paths to all the files in the provided directory.
	 * A given path will only be included once in the array; that is, there will be no duplicate paths.
	 * Note that this function excludes files that are deleted locally whether the deletion is staged or not.
	 *
	 * @param directory - A directory to filter the results by. Only files under this directory will be returned. To
	 * return all files in the repo use the value `"."`.
	 */
	public async getFiles(directory: string): Promise<string[]> {
		/**
		 * What these git ls-files flags do:
		 *
		 * ```
		 * --cached: Includes cached (staged) files.
		 * --others: Includes other (untracked) files that are not ignored.
		 * --exclude-standard: Excludes files that are ignored by standard ignore rules.
		 * --full-name: Shows the full path of the files relative to the repository root.
		 * ```
		 *
		 * Note that `--deduplicate` is not used here because it is not available until git version 2.31.0.
		 */
		const command = `ls-files --cached --others --exclude-standard --full-name ${directory}`;
		const [fileResults, deletedFiles] = await Promise.all([
			this.exec(command, "get files"),
			this.getDeletedFiles(),
		]);

		// Deduplicate the list of files by building a Set.
		// This includes paths to deleted, unstaged files, so we get the list of deleted files from git status and remove
		// those from the full list.
		const allFiles = new Set(
			fileResults
				.split("\n")
				.map((line) => line.trim())
				// filter out empty lines
				.filter((line) => line !== ""),
		);

		for (const deletedFile of deletedFiles) {
			allFiles.delete(deletedFile);
		}

		// Files are already repo root-relative
		return [...allFiles];
	}

	/**
	 * Execute git command
	 *
	 * @param command the git command
	 * @param error description of command line to print when error happens
	 */
	public async exec(command: string, error: string, pipeStdIn?: string) {
		return exec(`git ${command}`, this.resolvedRoot, error, pipeStdIn, {
			// Some git commands, like diff can have quite large output when there are very large changes like a pending merge with main.
			// To mitigate this, increase the maxBuffer size from its default (1 mb at time of writing).
			// https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback
			maxBuffer: 1024 * 1024 * 100,
		});
	}
}
