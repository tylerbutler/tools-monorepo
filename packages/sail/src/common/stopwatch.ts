import type { LoggingFunction } from "@tylerbu/cli-api";

/**
 * A stopwatch used for outputting messages to the terminal along with timestamp information.
 * The stopwatch is started upon creation, and each call to `log` will include the recorded time.
 */
export class Stopwatch {
	private lastTime: number = Date.now();
	private totalTime = 0;

	public constructor(
		private readonly enabled: boolean,
		protected logFunc: LoggingFunction,
	) {}

	public log(msg?: string, print?: boolean): number {
		const currTime = Date.now();
		const diffTime = currTime - this.lastTime;
		this.lastTime = currTime;
		const diffTimeInSeconds = diffTime / 1000;
		if (msg !== undefined) {
			if (this.enabled) {
				if (diffTime > 100) {
					this.logFunc(`${msg} - ${diffTimeInSeconds.toFixed(3)}s`);
				} else {
					this.logFunc(`${msg} - ${diffTime}ms`);
				}
			} else if (print === true) {
				this.logFunc(msg);
			}
		}
		this.totalTime += diffTime;
		return diffTimeInSeconds;
	}

	public getTotalTime(): number {
		return this.totalTime;
	}
}
