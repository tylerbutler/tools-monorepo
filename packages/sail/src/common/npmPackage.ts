import { existsSync } from "node:fs";
import path from "node:path";
import {
	type PackageJson as BasePackageJson,
	type IBuildProject,
	type IPackage,
	type IPackageManager,
	type IReleaseGroup,
	type IWorkspace,
	PackageBase,
	type ReleaseGroupName,
} from "@tylerbu/sail-infrastructure";
import registerDebug from "debug";

import type { ISailConfig } from "../core/sailConfig.js";
import { rimrafWithErrorAsync } from "./utils.js";

const traceInit = registerDebug("sail:init");

/**
 * A type representing Sail-specific config that may be in package.json.
 */
export type SailPackageJson = BasePackageJson & {
	/**
	 * Sail config. Some properties only apply when set in the root or release group root package.json.
	 */
	sail?: ISailConfig;

	/**
	 * For back-compat with fluid-build only.
	 *
	 * @deprecated Use the "sail" property instead.
	 */
	fluidBuild?: ISailConfig;
};

export class BuildPackage extends PackageBase<SailPackageJson> {
	private _matched = false;

	/**
	 * Create a new package from a package.json file. Prefer the .load method to calling the contructor directly.
	 *
	 * @param packageJsonFileName - The path to a package.json file.
	 * @param group - A group that this package is a part of.
	 * @param monoRepo - Set this if the package is part of a release group (monorepo).
	 * @param additionalProperties - An object with additional properties that should be added to the class. This is
	 * useful to augment the package class with additional properties.
	 */
	public constructor(packageInput: IPackage) {
		const {
			packageJsonFilePath,
			// packageManager,
			workspace,
			isWorkspaceRoot,
			releaseGroup,
			isReleaseGroupRoot,
		} = packageInput;
		super(
			packageJsonFilePath,
			// packageManager,
			workspace,
			isWorkspaceRoot,
			releaseGroup,
			isReleaseGroupRoot,
		);
		traceInit(`${this.nameColored}: Package loaded`);
		this.monoRepo = isWorkspaceRoot
			? new MonoRepo(
					releaseGroup,
					path.basename(packageJsonFilePath),
					releaseGroup,
					workspace,
				)
			: undefined;
	}

	public get matched() {
		return this._matched;
	}

	public set matched(value) {
		this._matched = value;
	}

	/**
	 * Get the full path for the lock file.
	 * @returns full path for the lock file, or undefined if one doesn't exist
	 */
	public getLockFilePath() {
		const directory = this.workspace.directory;
		for (const lockfile of this.workspace.packageManager.lockfileNames) {
			const full = path.join(directory, lockfile);
			if (existsSync(full)) {
				return full;
			}
		}
		return undefined;
	}

	public async cleanNodeModules() {
		return rimrafWithErrorAsync(
			path.join(this.directory, "node_modules"),
			this.nameColored,
		);
	}

	private _monoRepo: MonoRepo | undefined;

	private set monoRepo(value: MonoRepo | undefined) {
		this._monoRepo = value;
	}

	/**
	 * @deprecated Replace usage as soon as possible.
	 */
	public get monoRepo(): MonoRepo | undefined {
		return this._monoRepo;
	}
}

export class MonoRepo implements IWorkspace {
	public constructor(
		public readonly kind: string,
		public readonly repoPath: string,
		private readonly releaseGroupName: ReleaseGroupName,
		private readonly workspace: IWorkspace,
	) {}

	public get directory(): string {
		return this.workspace.directory;
	}

	public get packageManager(): IPackageManager {
		return this.workspace.packageManager;
	}

	public get rootPackage(): IPackage {
		return this.workspace.rootPackage;
	}

	public get releaseGroups(): Map<ReleaseGroupName, IReleaseGroup> {
		return this.workspace.releaseGroups;
	}

	public get buildProject(): IBuildProject<IPackage> {
		return this.workspace.buildProject;
	}

	public toString(): string {
		return this.workspace.toString();
	}

	public checkInstall(): Promise<true | string[]> {
		return this.workspace.checkInstall();
	}

	public install(updateLockfile: boolean): Promise<boolean> {
		return this.workspace.install(updateLockfile);
	}

	public reload(): void {
		this.workspace.reload();
	}

	public get name() {
		return this.workspace.name;
	}

	public get packages() {
		return this.workspace.packages;
	}

	private _releaseGroup: IReleaseGroup | undefined;

	public get releaseGroup() {
		if (this._releaseGroup === undefined) {
			this._releaseGroup = this.workspace.releaseGroups.get(
				this.releaseGroupName,
			);
			if (this._releaseGroup === undefined) {
				throw new Error(
					`Canot find release group "${this.releaseGroupName}" in workspace "${this.workspace.name}"`,
				);
			}
		}
		return this._releaseGroup;
	}

	public get version() {
		return this.releaseGroup.version;
	}
}
