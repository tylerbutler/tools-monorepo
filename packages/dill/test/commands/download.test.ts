import { readdir, rm } from "node:fs/promises";
import http from "node:http";
import { runCommand } from "@oclif/test";
import { getRandomPort } from "get-port-please";
import jsonfile from "jsonfile";
import path from "pathe";
import handler from "serve-handler";
import { temporaryDirectory } from "tempy";
import { withDir } from "tmp-promise";
import {
	afterAll,
	afterEach,
	beforeAll,
	beforeEach,
	describe,
	expect,
	it,
} from "vitest";

const { readFile: readJson } = jsonfile;

import process from "node:process";
import { getTestUrls, testDataPath } from "../common.js";

describe("download command", async () => {
	const server = http.createServer((request, response) => {
		// You pass two more arguments for config and middleware
		// More details here: https://github.com/vercel/serve-handler#options
		return handler(request, response, { public: testDataPath });
	});
	let port: number;
	let testUrls: URL[];

	beforeAll(async () => {
		port = await getRandomPort();
		testUrls = getTestUrls(port);
		server.listen(port);
	});

	let downloadDir: string;
	beforeEach(() => {
		downloadDir = temporaryDirectory();
	});

	afterEach(async () => {
		await rm(downloadDir, { recursive: true });
	});

	afterAll(() => {
		server.close();
	});

	it("downloads json", async () => {
		await runCommand(
			[
				// This is a single-command CLI, so use "." as the command entrypont per the oclif docs
				".",
				testUrls[0].toString(),
				`--out`,
				downloadDir,
			],
			{
				root: import.meta.url,
			},
		);

		const outputPath = path.join(downloadDir, "test0.json");
		const actual = await readJson(outputPath);
		expect(actual).to.deep.equal({
			key1: 1,
			key2: {
				nested: "object",
			},
		});
	});

	it("filename flag with --out", async () => {
		// process.chdir(downloadDir);
		const filename = "filename-flag-test.json";
		// path.join(downloadDir, "filename-flag-test.json");
		await runCommand(
			[
				// This is a single-command CLI, so use "." as the command entrypont per the oclif docs
				".",
				testUrls[0].toString(),
				`--out`,
				downloadDir,
				`--filename`,
				filename,
			],
			{
				root: import.meta.url,
			},
		);

		const outputPath = path.join(downloadDir, "filename-flag-test.json");
		const actual = await readJson(outputPath);
		expect(actual).to.deep.equal({
			key1: 1,
			key2: {
				nested: "object",
			},
		});
	});

	it("compressed file with --extract", async () => {
		await runCommand(
			[
				// This is a single-command CLI, so use "." as the command entrypont per the oclif docs
				".",
				testUrls[2].toString(),
				`--out`,
				downloadDir,
				"--extract",
			],
			{
				root: import.meta.url,
			},
		);

		const files = await readdir(downloadDir, { recursive: true });
		expect(files).toMatchSnapshot();
	});

	describe("using current working directory", () => {
		// let originalCwd: string;

		// beforeEach(() => {
		// 	originalCwd = process.cwd();
		// 	console.log(`original dir: ${originalCwd}`);
		// 	process.chdir(downloadDir);
		// 	console.log(`working dir: ${downloadDir}`);
		// });

		// afterEach(() => {
		// 	process.chdir(originalCwd);
		// });

		// biome-ignore lint/suspicious/noSkippedTests: feature doesn't work yet
		it.skip("with --filename", async () => {
			// const startingDir = process.cwd();
			await withDir(
				async ({ path: dlDir }) => {
					process.chdir(dlDir);
					const filename = "filename-flag-test.json";
					// path.join(downloadDir, "filename-flag-test.json");
					const { stdout } = await runCommand(
						[
							// This is a single-command CLI, so use "." as the command entrypont per the oclif docs
							".",
							testUrls[0].toString(),
							`--filename ${filename}`,
						],
						{
							root: import.meta.url,
						},
					);

					expect(stdout).to.contain("Downloading ");

					const outputPath = path.join(dlDir, "filename-flag-test.json");
					const actual = await readJson(outputPath);
					expect(actual).to.deep.equal({
						key1: 1,
						key2: {
							nested: "object",
						},
					});
				},
				{
					// usafeCleanup ensures the cleanup doesn't fail if there are files in the directory
					unsafeCleanup: true,
				},
			);
		});

		// it("--filename flag", async () => {
		// 	// process.chdir(downloadDir);
		// 	const filename = "filename-flag-test.json";
		// 	// path.join(downloadDir, "filename-flag-test.json");
		// 	const { stdout } = await runCommand(
		// 		[
		// 			// This is a single-command CLI, so use "." as the command entrypont per the oclif docs
		// 			".",
		// 			testUrls[0].toString(),
		// 			`--filename ${filename}`,
		// 		],
		// 		{
		// 			root: import.meta.url,
		// 		},
		// 	);

		// 	expect(stdout).to.contain("Downloading ");

		// 	const outputPath = path.join(downloadDir, "filename-flag-test.json");
		// 	const actual = await readJson(outputPath);
		// 	expect(actual).to.deep.equal({
		// 		key1: 1,
		// 		key2: {
		// 			nested: "object",
		// 		},
		// 	});
		// });
	});
});
