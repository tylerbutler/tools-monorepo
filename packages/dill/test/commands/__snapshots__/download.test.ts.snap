// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`download command > downloads json 1`] = `
{
  "data": [
    104,
    101,
    108,
    108,
    111,
    32,
    100,
    105,
    114,
    49,
  ],
  "type": "Buffer",
}
`;
